### PR TITLE
Modify name and version handling for logger collection

### LINKED ISSUE #9097: Checkpoints and loggers saving to different directories
## 🐛 Bug

The default `ModelCheckpoint()` callback is saving the checkpoint to a different directory than the directory for the `CSVLogger` logger I have set up. For example, when the first ever run is made the checkpoint is saved in `0/` while `hparams.yaml` and `metrics.csv` are saved in `version_0/`?

```
├── pl_logs
│   ├── 0/
│   │   ├── checkpoints/
│   ├── version_0/
│   │   ├── hparams.yaml
│       └── metrics.csv
```

### To Reproduce

```py
import pytorch_lightning as pl
from pytorch_lightning.callbacks.early_stopping import EarlyStopping
from pytorch_lightning.loggers import CSVLogger
from MyDataModule import MyDataModule
from MyModel import MyModel
from test_tube import HyperOptArgumentParser
from test_tube.hpc import SlurmCluster

def train(hparams, *args):
    ## Instantiate Modules ----
    dm = MyDataModule(batch_size=hparams.batch_size, nw=10)
    model = MyModel(
        lr=hparams.learning_rate,
        batch_size=hparams.batch_size,
    )

    ## Loggers ----
    log_csv = CSVLogger('.', name='pl_logs')
    loggers = [log_csv]

    ## Callbacks ----
    cb_earlystop = EarlyStopping(monitor='val_loss')
    callbacks = [cb_earlystop]

    ## Setup Trainer ----
    trainer = pl.Trainer(
        overfit_batches=10,
        min_epochs=1,
        max_epochs=1,
        num_sanity_val_steps=2,
        callbacks=callbacks,
        logger=loggers,
    )

    ## Train Model ----
    trainer.fit(model=model, datamodule=dm)

if __name__ == '__main__':
    ## Setup Hyperparameters ----
    parser = HyperOptArgumentParser(strategy='grid_search', add_help=False)
    parser.opt_list('--learning_rate', default=1e-3, type=float, options=[1e-4, 1e-3, 1e-2], tunable=True)
    parser.opt_list('--batch_size', default=2, type=int, options=[1, 2, 4, 6], tunable=True)
    hyperparams = parser.parse_args()

    ## Setup Slurm Cluster Object ----
    cluster = SlurmCluster(
        hyperparam_optimizer=hyperparams,
        log_path='/home/hunglab/mwarkent/projects/',
        python_cmd='python',
        enable_log_err=False
    )
    cluster.per_experiment_nb_cpus = 12
    cluster.per_experiment_nb_nodes = 1
    cluster.memory_mb_per_node = (16 * 1000)
    cluster.job_time = '07-00:00:00'

    ## Submit Jobs ----
    cluster.optimize_parallel_cluster_cpu(
        train_function=train,
        nb_trials=1,
        job_name='torch',
        job_display_name='torch'
    )
```

### Expected behavior

My expectation was that the checkpoints and all other logger files would be stored in the same versioned subdirectory. I recall this was happening when I used `pytorch_lightning` previously, but something seemingly has changed.

### Environment

- PyTorch Lightning Version: `pytorch-lightning 1.3.4`
- PyTorch Version: `torch 1.8.1`
- Python version: `Python 3.6.8`
- OS: Linux
- How you installed PyTorch: `pip`

cc @tchaton

### PR DESCRIPTION
**Pull request by mattwarkentin**
## What does this PR do?

Fixes #9097

This pull request is meant to simplify how `LoggerCollection` decides and shares `name` and `version` information when multiple loggers are used. When the loggers use the same version (the expected case), then only that version is returned (possibly as an integer) so that can be used by other machinery such as callbacks (e.g. `ModelCheckpoint`).

This should make it easier to get all of the items written to disk in a consistently versioned location for a given iteration/version of training.

### Does your PR introduce any breaking changes? If yes, please list them.

No breaking changes.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by mattwarkentin** [comment]
I'm fairly new to Python so this PR may not be ready in its current state. Happy to work on improving the PR, as needed.

File: pytorch_lightning/loggers/base.py [hunk 1]
```diff
@@ -410,18 +410,30 @@ def close(self) -> None:
         for logger in self._logger_iterable:
             logger.close()
 
+    @staticmethod
+    def all_equal(x):
+        return all(el == x[0] for el in x)
+
```
Comment (ananthsub) [inline pytorch_lightning/loggers/base.py:416]: 1. this doesn't need to be on the public API for the logger collection interface. we could define this as a helper outside the class and use it inside the `name` and `version` properties
2. please add typehints for new new code: https://github.com/PyTorchLightning/pytorch-lightning/issues/5023

i'm unclear what `x` is supposed to be and whether `x[0]` is a valid index access

**Review by ananthsub — COMMENTED** [review]
please add a unit test as well to demonstrate how the behavior changes with this additoin

File: pytorch_lightning/loggers/base.py [hunk 2]
```diff
@@ -410,18 +410,30 @@ def close(self) -> None:
         for logger in self._logger_iterable:
             logger.close()
 
+    @staticmethod
+    def all_equal(x):
+        return all(el == x[0] for el in x)
```
Comment (tchaton) [inline pytorch_lightning/loggers/base.py:415]:
```suggestion
    @staticmethod
    def _all_equal(x: str) -> bool:
        return all(el == x[0] for el in x)
```
